docs(inputs): revamp events articles (#1912)
telerik/blazor-docs: components/dateinput/events.md

This article explains the events available in the Telerik DateInput for Blazor:

* [OnBlur](#onblur)
* [OnChange](#onchange)
* [ValueChanged](#valuechanged)

## OnBlur

The `OnBlur` event fires when the component loses focus.

>caption Handle the OnBlur event

````CSHTML
@* You do not have to use OnChange to react to loss of focus *@

@result

<TelerikDateInput @bind-Value="@ТheInputValue"
                  OnBlur="@OnBlurHandler">
</TelerikDateInput>

@code{
    private string result = string.Empty;

    private DateTime? ТheInputValue { get; set; } = DateTime.Now;

    private void OnBlurHandler()
    {
        result = string.Format("BLUR fired, current value is {0:dd/MMM/yyyy}.", ТheInputValue);
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## OnChange

> The `OnChange` event represents an user action - confirmation of the current value. It fires when the user presses `Enter` in the input, or when the input loses focus.

Reviewer: ```suggestion
The `OnChange` event represents a user action - confirmation of the current value. It fires when the user presses `Enter` in the input, or when the input loses focus.
```

> The `OnChange` event represents a user action - confirmation of the current value. It fires when the user presses `Enter` in the input, or when the input loses focus.

Reviewer 2: Use an em dash instead of a hyphen when adding an explanation. The[ em dash rules](https://docs.telerik.com/style-guide/punctuation#lists) about lists also apply to paragraphs.
```suggestion
The `OnChange` event represents a user action—confirmation of the current value. It fires when the user presses `Enter` in the input or when the input loses focus.
```

> The DateInput is a generic component, so you must provide either a `Value`, or a type to the `T` parameter of the component.

Reviewer: ```suggestion
The DateInput is a generic component, so you must either provide a `Value`, or a type to the `T` parameter of the component.
```

This article explains the events available in the Telerik DateInput for Blazor:

* [OnBlur](#onblur)
* [OnChange](#onchange)
* [ValueChanged](#valuechanged)

## OnBlur

The `OnBlur` event fires when the component loses focus.

>caption Handle the OnBlur event

````CSHTML
@* You do not have to use OnChange to react to loss of focus *@

@result

<TelerikDateInput @bind-Value="@DateInputValue"
                  OnBlur="@OnBlurHandler">
</TelerikDateInput>

@code{
    private string result = string.Empty;

    private DateTime? DateInputValue { get; set; } = DateTime.Now;

    private void OnBlurHandler()
    {
        result = string.Format("BLUR fired, current value is {0:dd/MMM/yyyy}.", DateInputValue);
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## OnChange

> The `OnChange` event represents a user action that confirms the current value. It fires when the user presses `Enter` in the input or when the input loses focus.

Reviewer: ```suggestion
The `OnChange` event represents a user action that confirms the current value. It fires when the user presses `Enter` in the input or when the input loses focus.

The event handler receives an `object` argument that you need to cast to the actual `Value` type. The argument can hold a value or be `null`, depending on the user input and the `Value` type.
```

> 
> | Validation of input | Event handler argument | Nullable of not `Value` parameter |
> | --- | --- | --- |
> | Invalid input | `Null` | Nullable |
> | Invalid input | Default value of the `Value` parameter | Not nullable |
> | Valid input | The valid input as an object that needs to be cast to the type of the component | Nullable or not nullable |
> 

Reviewer: ```suggestion
```

The DateInput is a generic component, so you must either provide a `Value`, or a type to the `T` parameter of the component.

>caption Handle OnChange and use two-way binding

````CSHTML
@result
<br />
model value: @DateInputValue
<br />

<TelerikDateInput @bind-Value="@DateInputValue" 
                  OnChange="@MyOnChangeHandler">
</TelerikDateInput>

@code {
    private string result = string.Empty;

    private DateTime? DateInputValue { get; set; } = DateTime.Now;

    private void MyOnChangeHandler(object userInput)
    {
        // if you do not provide a Value, you must provide the Type parameter to the component
        result = string.Format("The user entered: {0:dd/MMM/yyyy}", (DateTime)userInput);
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

>tip The `OnChange` event is a custom event and does not interfere with bindings, so you can use it together with models and forms.

## ValueChanged

> The `ValueChanged` event fires on every typing in the input and if the input is valid. The handler will receive the valid input.

Reviewer: ```suggestion
The `ValueChanged` event fires:

 * During typing when the resulting input value is valid.
 * On input blur if the input value is not valid and the `Value` type is nullable.
```

> The event also fires if the input is invalid and when the input loses focus and the `Value` parameter is nullable. The handler will receive a `null` argument.

Reviewer: ```suggestion
```
